File: src/registry.ts
import type { ToolDefinition } from "./types.js";
import readFile from "./tools/read-file.js";
import writeFile from "./tools/write-file.js";
import runCommand from "./tools/run-command.js";
import gitStatus from "./tools/git-status.js";
import gitPush from "./tools/git-push.js";
import gitCommit from "./tools/git-commit.js";
import gitLog from "./tools/git-log.js";
import glob from "./tools/glob.js";
import grep from "./tools/grep.js";
import listDirectory from "./tools/list-directory.js";
import tree from "./tools/tree.js";
import applyPatch from "./tools/apply-patch.js";
import execCommand from "./tools/exec-command.js";
import writeStdin from "./tools/write-stdin.js";
import viewImage from "./tools/view-image.js";
import updatePlan from "./tools/update-plan.js";
import clockCurrTime from "./tools/clock-curr-time.js";
import clockSleep from "./tools/clock-sleep.js";
import getEnvironment from "./tools/get-environment.js";
import getProjectDoc from "./tools/get-project-doc.js";
import getAgentBrief from "./tools/get-agent-brief.js";

const ALL_TOOLS: ToolDefinition[] = [
  readFile,
  writeFile,
  runCommand,
  gitStatus,
  gitPush,
  gitCommit,
  gitLog,
  glob,
  grep,
  listDirectory,
  tree,
  // Ported from Codex (codex-rs/core/src/tools). Names use underscores because
  // MCP tool names must match ^[a-zA-Z0-9_-]{1,64}$, so Codex's `clock.curr_time`
  // becomes `clock_curr_time`.
  applyPatch,
  execCommand,
  writeStdin,
  viewImage,
  updatePlan,
  clockCurrTime,
  clockSleep,
  // None of these three is a Codex tool, because Codex does not need one: it
  // sends the OS and shell in an <environment_context> message, loads AGENTS.md
  // straight into the prompt, and puts its agent brief in the system prompt. An
  // MCP server has none of those channels, so each fact is offered as a tool
  // call as well as in the server's instructions. get_agent_brief returns all
  // three at once, for a client that only makes one onboarding call.
  getEnvironment,
  getProjectDoc,
  getAgentBrief,
];

export function loadTools(): ToolDefinition[] {
  const seen = new Set<string>();
  for (const tool of ALL_TOOLS) {
    if (seen.has(tool.name)) {
      throw new Error(`Duplicate tool name: ${tool.name}`);
    }
    seen.add(tool.name);
  }
  return ALL_TOOLS;
}

File: src/registry.rs
//! The tool registry. Ports `src/registry.ts`.
//!
//! `load_tools` returns every registered tool and enforces that names are
//! unique, panicking on a duplicate (a programming error, as in the TS which
//! throws at startup). The order mirrors the TypeScript registry.

use crate::tool::Tool;
use crate::tools;

pub fn load_tools() -> Vec<Box<dyn Tool>> {
    let all: Vec<Box<dyn Tool>> = vec![
        Box::new(tools::read_file::ReadFile),
        Box::new(tools::write_file::WriteFile),
        Box::new(tools::run_command::RunCommand),
        Box::new(tools::git_status::GitStatus),
        Box::new(tools::git_push::GitPush),
        Box::new(tools::git_commit::GitCommit),
        Box::new(tools::git_log::GitLog),
        Box::new(tools::glob::Glob),
        Box::new(tools::grep::Grep),
        Box::new(tools::list_directory::ListDirectory),
        Box::new(tools::tree::Tree),
        // Ported from Codex (codex-rs/core/src/tools). Names use underscores
        // because MCP tool names must match ^[a-zA-Z0-9_-]{1,64}$.
        Box::new(tools::apply_patch::ApplyPatch),
        Box::new(tools::exec_command::ExecCommand),
        Box::new(tools::write_stdin::WriteStdin),
        Box::new(tools::view_image::ViewImage),
        Box::new(tools::update_plan::UpdatePlan),
        Box::new(tools::clock_curr_time::ClockCurrTime),
        Box::new(tools::clock_sleep::ClockSleep),
        // None of these three is a Codex tool: get_environment, get_project_doc
        // and get_agent_brief surface facts Codex sends through channels an MCP
        // server does not have.
        Box::new(tools::get_environment::GetEnvironment),
        Box::new(tools::get_project_doc::GetProjectDoc),
        Box::new(tools::get_agent_brief::GetAgentBrief),
        // Persistent working memory: what a chat window loses between conversations.
        Box::new(tools::remember::Remember),
        Box::new(tools::recall::Recall),
        // Codex's skills.list / skills.read.
        Box::new(tools::skills_list::SkillsList),
        Box::new(tools::skills_read::SkillsRead),
    ];

    let mut seen = std::collections::HashSet::new();
    for tool in &all {
        if !seen.insert(tool.name()) {
            panic!("Duplicate tool name: {}", tool.name());
        }
    }
    all
}
